[READY] cool laser gun update

Tlaltecuhtli opened the pull request:
:cl:
imageadd: changed the laser gun sprite to look more compact
add: external lens: they add a various special ammo types to standard laser guns when hit with it, they are available by research or cargo
/:cl:

laser guns are boring so i thought of spicing them up and shake up the meta by adding new available ammo types which do unique things:

- shock: it applies shock (no stun) and has a chance to make the gun drop, good vs nanite users, its projectiles consume more energy 

- blinding: chance to blind the dude for 5s + some eye blur high energy use and does low dmg

- bouncing: it bounces on walls usefull for trick shooting and maint sweeping higher energy use and does lower dmg

- stealth: almost invisible projectile it costs more to fire and low range

- incendiary: chance to set him on fire


- economic: costs less power and does less damage

- shield:  projects an holo cover for a few seconds

- scatter: it shoots 5 10 dmg mini lasers with a wide spread



to do list:

- [x] add sounds

- [x] better lens sprites

- [x] fix problems

- [x] add more details and fluff

- [x] balancing 

- [x]  add more stuff if it sounds cool

- [x] fix conflicts

Tlaltecuhtli committed:
1

Tlaltecuhtli committed:
1

Tlaltecuhtli committed:
2

Tlaltecuhtli committed:
2

Tlaltecuhtli committed:
3

Tlaltecuhtli committed:
3

ATH1909 commented on code/modules/projectiles/projectile/beams.dm:
```diff
@@ -184,3 +184,232 @@
+			var/mob/living/carbon/human/FH = firer
+			var/obj/item/card/id/FC = FH.get_idcard(TRUE)
+			if(FC && FC.registered_account)
+				FC.registered_account.adjust_money(100)
```
shouldn't this only add however much money you stole to your account?

also, does this even check if the target's account has any money in it?

ATH1909 commented on code/modules/projectiles/projectile/beams.dm:
```diff
@@ -184,3 +184,232 @@
+
+/obj/screen/alert/status_effect/laserweak
+	name = "high laser absortion"
+	desc = "You are covered in an high laser absorbing gel"
```
"an high laser"

zxaber commented:
Not really a huge fan of the ticket machine. Memephone aside, whoever controls an ID card should get to choose when money is pulled out. If security wants to fine people, they can just pull their ID off and take the money directly. I guarantee this will primarily be used for griefing.

Invisible lasers are worth a lot of thought. We're talking about allowing people to take damage with no clue as to the source.

Heavy lasers breaking walls makes this the gun form of hulk. Having people just randomly delete walls in your department because they can is never fun.

The other mods sound neat. We should have them exclusively be cargo items.

Shapsy commented:
The ticketing one sounds really dumb but also admittedly pretty funny, and would be neat to see as an alt firemode for sec's disablers

WJohn commented:
holy shit these sound insane

OverlyGeneric commented:
Tracing makes people highly-irradiated which gives that lovely glow.

Michael-Ashfield commented:
Could these be a late stage research node?
To be honest gun science, with upgradable powercells and components sounds like a pretty fun science that may actually give a use to the currently useless science range, it would also allow science and security to work together to make crazy weapons.

Tlaltecuhtli commented:
it is in late stage research already

habader commented:
no please god no

OnlineGirlfriend commented:
> * heavy: breaks normal walls, irradiates, does 50% extra dmg and some armor pen but it has really low ammo count

Is the ammo count 1?  because that’s the only way I see this being balanced.

ninjanomnom commented on code/modules/projectiles/guns/energy/laser.dm:
```diff
@@ -138,3 +164,69 @@
+
+/datum/component/Scharge/Initialize()
+	. = ..()
+	RegisterSignal(parent,"syphon", .proc/charge)
```
I know this is WIP but please make sure to follow the standards of components and signals. 

Signals must use defines to avoid magic strings like this.

Components must be in the components folder.

ninjanomnom commented on code/modules/projectiles/guns/energy/laser.dm:
```diff
@@ -138,3 +164,69 @@
+	. = ..()
+	RegisterSignal(parent,"syphon", .proc/charge)
+
+/datum/component/Scharge/proc/charge(mob/FH) //doenst work
```
The first argument is always the signal source as documented in the signals defines file.

PKPenguin321 commented:
does incendiary have any downside

AnturK commented on code/modules/cargo/packs.dm:
```diff
@@ -303,6 +303,28 @@
+	name = "Assorted external lens Crate"
+	desc = "Contains five experimental lens that can be applied to standard laser guns. Requires Security access to open."
+	cost = 8000
+	contains = list(/obj/item/external_lens/BTC,
```
Don't use uppercase in paths, also give these thing names that actually mean something.

Detective-Google commented:
```
/obj/item/external_lens
	name = "external lens: scatter ray"
	icon_state = "external_scatter"
	desc = "external lens."
	icon = 'icons/obj/guns/energy.dmi'
	var/stored_ammo_type = /obj/item/ammo_casing/energy/laser/scatter
```
This isn't mentioned on the main body of the PR, is it? Scatter lasers are super powerful, and you make no mention as to the balance of it, are you trying to "sneak" this in? or what?

`/datum/design/lensSCT
	name = "external lens: scatter"
	desc = "external lens"
	id = "lens_SCT"
	build_type = PROTOLATHE
	materials = list(MAT_METAL = 2000, MAT_GLASS = 6000, MAT_URANIUM = 1500, MAT_SILVER = 2000,  MAT_GOLD = 5000)
	build_path = /obj/item/external_lens
	category = list("Ammo")
	departmental_flags = DEPARTMENTAL_FLAG_SECURITY`

this lens is mentioned literally no where on your PR, so unless I'm missing the purpose of this code, it seems like you're trying to add this without telling anyone. I did a control F and nowhere on this PR is the word "scatter" mentioned.

AnturK commented on code/modules/projectiles/guns/energy/laser.dm:
```diff
@@ -8,6 +8,32 @@
 	shaded_charge = 1
+	var/obj/item/external_lens/stored = null
+
+/obj/item/gun/energy/laser/attackby(obj/item/I, mob/user, params)
```
Given this is on base path with no checks, this allows lasertag/practice guns and other weird subtypes to be turned into fully functional guns.

optimumtact commented:
you are actually the new goof it's sad.

I should gitban you now

necromanceranne commented:
I'm pretty sure the way you want tracers to work, it makes you vulnerable to all burn damage and apparently stun? I don't know why you included that but it's worth mentioning that if you also added incendiary which benefits both from the fire stacks dealing burn damage and the lasers themselves. Couldn't you just have it reduce their laser armor?

Tlaltecuhtli committed:
gfgddf

Tlaltecuhtli committed:
gfgddf

Tlaltecuhtli committed:
dfsdfsf

Tlaltecuhtli committed:
dfsdfsf

Tlaltecuhtli committed:
dfsf

Tlaltecuhtli committed:
dfsf

FantasticFwoosh commented:
I like this for making laser guns more viable, as often people skip straight to shotguns or autorifles for their exotic specialist ammotypes. Doesn't require making multiple types of the same gun and has a lot of both fun and practical utilities :+1: 

I will say that I dont think the ticket dispensing lens is very tasteful, with all the code you have there you could make a seperate tool object to do it instead. Id lap that right up and use that all the time rather than virtually mug misbehaving assistants for ID they rarely carry anyway.
> keep the stamina damage too, that made me chuckle.

Tlaltecuhtli committed:
shield + lens icons (untested)

Tlaltecuhtli committed:
shield + lens icons (untested)

ninjanomnom commented on code/modules/projectiles/guns/energy/laser.dm:
```diff
@@ -8,6 +8,33 @@
+	..()
+	if(istype(I, /obj/item/external_lens) && modifystate)
+		if(!stored)
+			equiplens(I,user)
```
You should really make the lenses apply a component instead of this

Tlaltecuhtli replied:
@ninjanomnom i made it a component but i m not sure how to remove the ammo type when detaching the lens, as its a list of proj objects and if i qdel the proj it creates a null which breaks selection 
are there things that remove from a list and also fix the list lenght?

Tlaltecuhtli committed:
component

Tlaltecuhtli committed:
component

Tlaltecuhtli committed:
now it works

Tlaltecuhtli committed:
now it works

ninjanomnom commented on code/modules/projectiles/guns/energy/laser.dm:
```diff
@@ -241,5 +232,26 @@
+	stored_ammo_type = /obj/item/ammo_casing/energy/laser/shield
+	overlay = "shield"
+
+/datum/component/extralasers //will move it to another file when it works
```
This needs to be in the component folder

ninjanomnom commented on code/modules/projectiles/guns/energy/laser.dm:
```diff
@@ -241,5 +232,26 @@
+
+/datum/component/extralasers //will move it to another file when it works
+	var/obj/item/external_lens/lens
+/datum/component/extralasers/Initialize()
```
Please put some whitespace before procs/types to make things a little more readable. It's also how we do it most everywhere anyway.

ninjanomnom replied:
Haven't I told you this before

ninjanomnom commented on code/modules/projectiles/guns/energy/laser.dm:
```diff
@@ -8,6 +8,17 @@
+	modifystate = TRUE
+
+/obj/item/gun/energy/laser/Initialize()
+	..()
```
```suggestion
	. = ..()
```

ninjanomnom commented on code/modules/projectiles/guns/energy/laser.dm:
```diff
@@ -8,6 +8,17 @@
+	AddComponent(/datum/component/extralasers)
+
+/obj/item/gun/energy/laser/attackby(obj/item/I, mob/user, params)
+	..()
```
```suggestion
	. = ..()
```

ninjanomnom commented on code/modules/projectiles/guns/energy/laser.dm:
```diff
@@ -8,6 +8,17 @@
+/obj/item/gun/energy/laser/attackby(obj/item/I, mob/user, params)
+	..()
+	if(modifystate && istype(I, /obj/item/crowbar))
+		SEND_SIGNAL(src,COMSIG_DETACH_LENS)
```
Use the attackby signal instead

ninjanomnom commented on code/modules/projectiles/guns/energy/laser.dm:
```diff
@@ -138,3 +149,112 @@
+	var/overlay = "laser"
+
+/obj/item/external_lens/Initialize()
+	..()
```
```suggestion
	. = ..()
```

ninjanomnom commented on code/modules/projectiles/guns/energy/laser.dm:
```diff
@@ -138,3 +149,112 @@
+
+/obj/item/external_lens/Initialize()
+	..()
+	add_overlay("[overlay]")
```
```suggestion
	add_overlay(overlay)
```

ninjanomnom commented on code/modules/projectiles/guns/energy/laser.dm:
```diff
@@ -138,3 +149,112 @@
+/obj/item/external_lens/afterattack(atom/movable/AM, mob/user, flag)
+	. = ..()
+	if(istype(AM, /obj/item/gun/energy/laser))
+		SEND_SIGNAL(AM,COMSIG_ATTACH_LENS, stored_ammo_type,src)
```
Use the afterattack signal instead

ninjanomnom commented on code/modules/projectiles/guns/energy/laser.dm:
```diff
@@ -8,6 +8,17 @@
+
+/obj/item/gun/energy/laser/Initialize()
+	..()
+	AddComponent(/datum/component/extralasers)
```
The component should be added by the thing doing the modification

Tlaltecuhtli committed:
component2

Tlaltecuhtli committed:
component2

ninjanomnom commented on code/modules/projectiles/guns/energy/laser.dm:
```diff
@@ -10,14 +10,10 @@
+	. = ..()
 	if(modifystate && istype(I, /obj/item/crowbar))
-		SEND_SIGNAL(src,COMSIG_DETACH_LENS)
+		SEND_SIGNAL(src, COMSIG_PARENT_ATTACKBY)
```
We already have an attackby signal, you need to use that. Do not make a new signal here of any kind

Tlaltecuhtli committed:
bugfix

Tlaltecuhtli committed:
bugfix

Tlaltecuhtli committed:
stuff

Tlaltecuhtli committed:
stuff

Tlaltecuhtli committed:
crat

Tlaltecuhtli committed:
crat

ninjanomnom commented on code/datums/components/laser_components.dm:
```diff
@@ -0,0 +1,25 @@
+	lens = _lens
+	RegisterSignal(parent, COMSIG_PARENT_ATTACKBY, .proc/detach)
+
+	var/obj/item/ammo_casing/energy/laser/shoot =  ammo
```
Just make it a typed argument

ninjanomnom commented on code/datums/components/laser_components.dm:
```diff
@@ -0,0 +1,25 @@
+	RegisterSignal(parent, COMSIG_PARENT_ATTACKBY, .proc/detach)
+
+	var/obj/item/ammo_casing/energy/laser/shoot =  ammo
+	var/obj/item/gun/energy/laser/L = parent
```
You need to `return COMPONENT_INCOMPATIBLE` if parent isnt the correct type

ninjanomnom commented on code/game/machinery/deployable.dm:
```diff
@@ -191,6 +191,27 @@
+	icon_state = "holobarrier1"
+
+/obj/structure/barricade/security/holobarricade/Initialize()
+	..()
```
```suggestion
	. = ..()
```

ninjanomnom commented on code/game/machinery/deployable.dm:
```diff
@@ -191,6 +191,27 @@
+	QDEL_IN(src, 120)
+
+/obj/structure/barricade/security/holobarricade/attackby(obj/item/I, mob/user, params)
+	return ..()
```
???

ninjanomnom commented on code/modules/projectiles/projectile.dm:
```diff
@@ -223,16 +223,17 @@
-			if(hitscan)
-				store_hitscan_collision(pcache)
-			return TRUE
+	if(ricochets < ricochets_max)
```
Early return on this if check

ninjanomnom commented on code/modules/projectiles/guns/energy/laser.dm:
```diff
@@ -138,3 +139,95 @@
+		var/obj/item/gun/energy/laser/L = AM
+		if(!L.modifystate)
+			L.AddComponent(/datum/component/extralasers, stored_ammo_type, src)
+			forceMove(L)
```
Make the lens get deleted when applying it and make a new one from the component when it's removed

ninjanomnom replied:
Don't resolve things that aren't resolved

ninjanomnom commented on code/modules/projectiles/guns/energy/laser.dm:
```diff
@@ -138,3 +139,95 @@
+	. = ..()
+	if(istype(AM, /obj/item/gun/energy/laser))
+		var/obj/item/gun/energy/laser/L = AM
+		if(!L.modifystate)
```
Please do not reuse this variable for this purpose. Use the component tools to handle what happens when multiple of the component is added to a gun

Tlaltecuhtli replied:
i was using getcomponent to check if it had the laser component already, which works but this happens on something i havent touched http://take.ms/iO1AD

ninjanomnom replied:
That error is unrelated to anything you're doing here

ninjanomnom commented on code/modules/projectiles/guns/energy/laser.dm:
```diff
@@ -138,3 +139,95 @@
+			L.AddComponent(/datum/component/extralasers, stored_ammo_type, src)
+			forceMove(L)
+			L.modifystate = TRUE
+			return
```
You do not need to return on the last line ever

ninjanomnom commented on code/modules/projectiles/projectile/beams.dm:
```diff
@@ -184,3 +184,150 @@
+
+/obj/item/projectile/beam/bitcoin/on_hit(atom/target)
+	. = ..()
+	if(ishuman(target))
```
Early return on this if check

ninjanomnom commented on code/modules/projectiles/projectile/beams.dm:
```diff
@@ -184,3 +184,150 @@
+	if(ishuman(target))
+		var/mob/living/carbon/human/H = target
+		var/obj/item/card/id/C = H.get_idcard(TRUE)
+		if(C && C.registered_account && !(C.registered_account.account_balance < 50))
```
```suggestion
		if(C?.registered_account && !(C.registered_account.account_balance < 50))
```

ninjanomnom commented on code/modules/projectiles/projectile/beams.dm:
```diff
@@ -184,3 +184,150 @@
+			if(ishuman(firer))
+				var/mob/living/carbon/human/FH = firer
+				var/obj/item/card/id/FC = FH.get_idcard(TRUE)
+				if(FC && FC.registered_account)
```
```suggestion
				if(FC?.registered_account)
```

ninjanomnom commented on code/modules/projectiles/projectile/beams.dm:
```diff
@@ -184,3 +184,150 @@
+			to_chat(H, "<span class='danger'>Your wallet feels lighter!</span>")
+			if(ishuman(firer))
+				var/mob/living/carbon/human/FH = firer
+				var/obj/item/card/id/FC = FH.get_idcard(TRUE)
```
Use some better names

ninjanomnom commented on code/modules/projectiles/projectile/beams.dm:
```diff
@@ -184,3 +184,150 @@
+			else
+				var/datum/bank_account/D = SSeconomy.get_dep_account(ACCOUNT_SEC)
+				if(D)
+					D.adjust_money(50)
```
remove the if
`D?.adjust_money(50)`

ninjanomnom commented on code/modules/projectiles/projectile/beams.dm:
```diff
@@ -184,3 +184,150 @@
+					D.adjust_money(50)
+		else
+			H.adjustStaminaLoss(40)
+			to_chat(H, "<span class='danger'>You feel your energy itself being turned into cryptocurrency!</span>")
```
This message isn't even true anymore and making it true seems like a very easily abusable feature

4dplanner replied:
40 stamloss is too high under the new system, needs adjustment

ninjanomnom commented on code/modules/projectiles/projectile/energy/net_snare.dm:
```diff
@@ -95,3 +95,24 @@
+	if(!locate(/obj/structure/barricade/security/holobarricade) in Tloc)
+		new /obj/structure/barricade/security/holobarricade(Tloc)
+	do_sparks(1, TRUE, src)
+	qdel(src)
```
```suggestion
	return ..()
```

ninjanomnom commented:
Making all of these available via cargo crate seems like a poor idea

Tlaltecuhtli committed:
Update code/modules/projectiles/projectile/energy/net_snare.dm

Co-Authored-By: Emmett Gaines <ninjanomnom@gmail.com>

Tlaltecuhtli committed:
Update code/modules/projectiles/projectile/energy/net_snare.dm

Co-Authored-By: Emmett Gaines <ninjanomnom@gmail.com>

Tlaltecuhtli committed:
Update code/modules/projectiles/projectile/beams.dm

Co-Authored-By: Emmett Gaines <ninjanomnom@gmail.com>

Tlaltecuhtli committed:
Update code/modules/projectiles/projectile/beams.dm

Co-Authored-By: Emmett Gaines <ninjanomnom@gmail.com>

Tlaltecuhtli committed:
Update code/game/machinery/deployable.dm

Co-Authored-By: Emmett Gaines <ninjanomnom@gmail.com>

Tlaltecuhtli committed:
Update code/game/machinery/deployable.dm

Co-Authored-By: Emmett Gaines <ninjanomnom@gmail.com>

Tlaltecuhtli committed:
Update code/modules/projectiles/projectile/beams.dm

Co-Authored-By: Emmett Gaines <ninjanomnom@gmail.com>

Tlaltecuhtli committed:
Update code/modules/projectiles/projectile/beams.dm

Co-Authored-By: Emmett Gaines <ninjanomnom@gmail.com>

Tlaltecuhtli committed:
d

Tlaltecuhtli committed:
d

ninjanomnom commented on code/modules/projectiles/projectile.dm:
```diff
@@ -223,16 +223,19 @@
+					store_hitscan_collision(pcache)
+				return TRUE
+	else
+		return
```
What is this

ninjanomnom commented on code/datums/components/laser_components.dm:
```diff
@@ -0,0 +1,26 @@
+		L.select_fire(user)
+		L.recharge_newshot()
+		L.update_icon(TRUE)
+		RemoveComponent()
```
qdel is fine

ninjanomnom commented on code/modules/projectiles/guns/energy/laser.dm:
```diff
@@ -138,3 +139,93 @@
+	. = ..()
+	if(istype(AM, /obj/item/gun/energy/laser))
+		var/obj/item/gun/energy/laser/L = AM
+		if(!L.GetComponent(/datum/component/extralasers))
```
You do not need to GetComponent here, if you're unsure how to use components contact me with questions somewhere

Tlaltecuhtli replied:
what do i need then

ninjanomnom replied:
I said earlier, you can make use of how components inherit to make all this get handled in the component

Tlaltecuhtli replied:
do i have to create the component on initialize inside the lens and then move it into the gun?

ninjanomnom replied:
No, components have special handling for what happens if you apply a component to something that already has that kind of component. You can make use of that to handle what happens when you try to apply a lens on top of another lens

Tlaltecuhtli committed:
merge conflict

Tlaltecuhtli committed:
merge conflict

Tlaltecuhtli committed:
early return

Tlaltecuhtli committed:
early return

ninjanomnom commented on code/modules/projectiles/projectile.dm:
```diff
@@ -223,7 +223,9 @@
 	var/datum/point/pcache = trajectory.copy_to()
 	var/turf/T = get_turf(A)
-	if(check_ricochet(A) && check_ricochet_flag(A) && ricochets < ricochets_max)
+	if(ricochets >= ricochets_max)
```
Sorry, I made a mistake reading this earlier, rather: why was this separated out into a separate if check?

Tlaltecuhtli replied:
first check is required for both ricochet and normal projectiles, then it does an OR to see if it respects the two flags or its a ricochet proj

ninjanomnom replied:
No, why is it in a seperate if statement? This was part of the if statement below originally and you split it out

Tlaltecuhtli committed:
sdasd

Tlaltecuhtli committed:
sdasd

Tlaltecuhtli commented on code/datums/components/laser_components.dm:
```diff
@@ -0,0 +1,27 @@
+		L.update_icon(TRUE)
+		RemoveComponent()
+		var/turf/T = user.loc
+		new ammo(T) //doest work
```
@ninjanomnom for some reason it gives the error "cant create object of obj/item/external_lens/type"

ninjanomnom replied:
Doesn't look like you ever assign to the ammo var

Tlaltecuhtli replied:
uhh i think the problem was lens was a reference to a deleted object and new wants a path to the object
this new thing gives the runtime that it fired COMPONENT INCOMPATIBLE on  the mob and idk why

ninjanomnom replied:
Because you're trying to apply the component to the wrong type

Tlaltecuhtli committed:
sprite fixes

Tlaltecuhtli committed:
sprite fixes

Tlaltecuhtli committed:
weird stuff

Tlaltecuhtli committed:
weird stuff

ninjanomnom commented on code/datums/components/laser_components.dm:
```diff
@@ -0,0 +1,28 @@
+	path = text2path( "/obj/item/external_lens/[overlay]")
+	RegisterSignal(parent, COMSIG_PARENT_ATTACKBY, .proc/detach)
+
+/datum/component/extralasers/proc/detach(datum/source, obj/item/I, mob/user, params)
```
You need to make component inheritance call on this

Tlaltecuhtli replied:
what

Tlaltecuhtli commented on code/datums/components/laser_components.dm:
```diff
@@ -0,0 +1,28 @@
+	L.ammo_type  += new ammo (src)
+
+	qdel(_lens)
+	path = text2path( "/obj/item/external_lens/[overlay]")
```
is there a function to get the path of the item?

ninjanomnom replied:
`item.type`

Tlaltecuhtli committed:
fix

Tlaltecuhtli committed:
fix

Tlaltecuhtli committed:
?

Tlaltecuhtli committed:
?

optimumtact commented:
@tgstation/commit-access anyone particularly want to work on this for balance and fixes? I don't

ninjanomnom commented:
Gameplay wise I don't care for this very much, I just want the code to be good if we are going to have it

Tlaltecuhtli committed:
fix

Tlaltecuhtli committed:
fix

Tlaltecuhtli committed:
wsdasd

Tlaltecuhtli committed:
wsdasd

ExcessiveUseOfCobblestone commented:
I don't want to balance this lmao

PKPenguin321 commented:
Just merge it it'll be fine

Tlaltecuhtli commented:
it seems balanced to me, if something really wrong happens ingame later i can just tweak the numbers

optimumtact commented:
Im going to leave this open for 24 hours and if nobody from @tgstation/commit-access is willing to touch it for final merge I'll close it.

FloranOtten commented:
[Dawn on the First Day]
   {19 hours remain}

ExcessiveUseOfCobblestone commented:
i'll entertain shield/eco/fire/stealth/bouncy/blind and maybe shock.

Also 20 days and scatter is still undoc'd.

Tlaltecuhtli commented:
@ExcessiveUseOfCobblestone scatter is now documented

ExcessiveUseOfCobblestone commented:
ill entertain scatter too, let me know when just the ones I want are in the code.

btw you should subtype the projectiles to laser.

Tlaltecuhtli committed:
rip bitcoins

Tlaltecuhtli committed:
rip bitcoins

Tlaltecuhtli committed:
e

Tlaltecuhtli committed:
e

Tlaltecuhtli committed:
e

Tlaltecuhtli committed:
e

Tlaltecuhtli committed:
laser subtype

Tlaltecuhtli committed:
laser subtype

Tlaltecuhtli committed:
wew

Tlaltecuhtli committed:
wew

ExcessiveUseOfCobblestone commented:
cringe why did I say ill entertain a tlat pr

Tlaltecuhtli committed:
e

Tlaltecuhtli committed:
e

Tlaltecuhtli commented:
hey my compiler broke so i m stuck on doing small changes on webeditor until i find how to fix byond, the code works 💯 % fine 👌

ninjanomnom commented:
How are you supposed to know the code works when you cant even compile it to test

ninjanomnom commented on code/datums/components/laser_components.dm:
```diff
@@ -0,0 +1,32 @@
+		L.update_icon(TRUE)
+		var/turf/T = user.loc
+		new path(T)
+		RemoveComponent()
```
qdel

ninjanomnom commented on code/datums/components/laser_components.dm:
```diff
@@ -0,0 +1,32 @@
+/datum/component/extralasers/proc/attackby(datum/source, obj/item/I, mob/user, params)
+	if(I.tool_behaviour == TOOL_CROWBAR)
+		var/obj/item/gun/energy/laser/L = parent
+		L?.chambered = null
```
If you dont have a parent something horrible has gone wrong

ninjanomnom commented on code/datums/components/laser_components.dm:
```diff
@@ -0,0 +1,32 @@
+		var/obj/item/gun/energy/laser/L = parent
+		L?.chambered = null
+		if(L.ammo_type)
+			var/obj/item/ammo_casing/energy/R = LAZYACCESS(L.ammo_type, L.ammo_type.len)
```
How do you know your ammo type is at the end? You should be keeping track of what you added to remove it properly

ninjanomnom commented on code/datums/components/laser_components.dm:
```diff
@@ -0,0 +1,32 @@
+/datum/component/extralasers
+	dupe_mode = COMPONENT_DUPE_UNIQUE_PASSARGS
+	var/path
```
choose a better name

ninjanomnom commented on code/datums/components/laser_components.dm:
```diff
@@ -0,0 +1,32 @@
+/datum/component/extralasers
+	dupe_mode = COMPONENT_DUPE_UNIQUE_PASSARGS
```
You dont even use the inherit proc

ninjanomnom commented on code/datums/components/laser_components.dm:
```diff
@@ -0,0 +1,32 @@
+	var/obj/item/gun/energy/laser/L = parent
+	L.ammo_type  += new ammo (src)
+	path = _path
+	qdel(_lens)
```
You don't do anything with the lens besides delete it

ninjanomnom commented on code/datums/components/laser_components.dm:
```diff
@@ -0,0 +1,32 @@
+		var/turf/T = user.loc
+		new path(T)
+		RemoveComponent()
+		return TRUE
```
It's a flag, not a bool. Use the correct define

Tlaltecuhtli committed:
c

Tlaltecuhtli committed:
c

Tlaltecuhtli committed:
plan :b:

Tlaltecuhtli committed:
plan :b:

Tlaltecuhtli committed:
epic

Tlaltecuhtli committed:
epic

Tlaltecuhtli commented:
@ninjanomnom is this way ok? or having dupe_higlander with edited qdel proc to detach?

ninjanomnom commented:
As I keep saying there is a proc that gets called during component inheritance, make use of that

Tlaltecuhtli committed:
Update laser_components.dm

Tlaltecuhtli committed:
Update laser_components.dm

ninjanomnom commented on code/datums/components/laser_components.dm:
```diff
@@ -0,0 +1,47 @@
+
+/datum/component/extralasers/InheritComponent(datum/newcomp, orig, list/arglist)
+	. = ..()
+	detach(0, user)
```
```suggestion
	detach(null, user)
```

ninjanomnom commented on code/datums/components/laser_components.dm:
```diff
@@ -0,0 +1,47 @@
+	qdel(lens)
+
+/datum/component/extralasers/proc/detach(datum/source, mob/user, breaklens = FALSE)
+	if(parent)
```
You shouldn't need to check if you have a parent here

ninjanomnom commented on code/datums/components/laser_components.dm:
```diff
@@ -0,0 +1,47 @@
+	ammo =  new _ammo (src)
+	L.ammo_type  += ammo
+	lens_path = _lens_path
+	qdel(lens)
```
You're still not doing anything with this lens besides deleteing it, why are you passing it in?

ninjanomnom commented on code/datums/components/laser_components.dm:
```diff
@@ -0,0 +1,47 @@
+/datum/component/extralasers/proc/attach(obj/item/ammo_casing/energy/_ammo, _lens_path, lens)
+	var/obj/item/gun/energy/laser/L = parent
+	ammo =  new _ammo (src)
+	L.ammo_type  += ammo
```
Isn't this supposed to be types not instances?

Tlaltecuhtli replied:
ammo_type is made of instances

ninjanomnom replied:
I just looked at it's types

Tlaltecuhtli replied:
https://github.com/tgstation/tgstation/blob/a8fa220f7fd601f206363f68fff580b731edc0d4/code/modules/projectiles/guns/energy.dm#L53

ninjanomnom replied:
Ugh, why are only energy weapons storing instances

Tlaltecuhtli committed:
clean up ( sprites got delete rip reeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee)

Tlaltecuhtli committed:
clean up ( sprites got delete rip reeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee)

Tlaltecuhtli committed:
srit1

Tlaltecuhtli committed:
srit1

Tlaltecuhtli committed:
2

Tlaltecuhtli committed:
2

Tlaltecuhtli committed:
screams

Tlaltecuhtli committed:
screams

Tlaltecuhtli committed:
no parent?

Tlaltecuhtli committed:
no parent?

ninjanomnom commented on code/datums/components/laser_components.dm:
```diff
@@ -0,0 +1,44 @@
+
+/datum/component/extralasers/proc/detach(breaklens = FALSE)
+	var/obj/item/gun/energy/laser/L = parent
+	L?.chambered = null
```
You're checking again if parent exists here, this should be unnecesary

ninjanomnom commented on code/datums/components/laser_components.dm:
```diff
@@ -0,0 +1,44 @@
+	var/obj/item/gun/energy/laser/L = parent
+	L?.chambered = null
+	if(L.ammo_type.len)
+		LAZYREMOVE(L.ammo_type, ammo)
```
You don't need the if here, just use LAZYREMOVE directly

Tlaltecuhtli committed:
d

Tlaltecuhtli committed:
d

Tlaltecuhtli committed:
exact laser type

Tlaltecuhtli committed:
exact laser type

Tlaltecuhtli committed:
e

Tlaltecuhtli committed:
e

optimumtact commented:
remove the scatter module

Tlaltecuhtli committed:
removes scatter

Tlaltecuhtli committed:
e

Tlaltecuhtli committed:
e

ExcessiveUseOfCobblestone commented:
I'll try to do a balance runthru this evening so this can get off my precious tracker.

ExcessiveUseOfCobblestone commented:
These look alright from a quick balance review, my only concern is that you can bypass the techweb requirements entirely if you use cargo (and I'm not sure how well the sec budget is balanced).

Tlaltecuhtli commented:
sec gets money from: 1500 * % of crew alive (lol), killing monsters in the station z level (lol) and gulag money(lol)

ninjanomnom commented on code/modules/projectiles/guns/energy/laser.dm:
```diff
@@ -138,3 +139,68 @@
+
+//////////////EXTERNAL LENS////////////////////////////
+
+/obj/item/external_lens
```
Put these in their own file

ninjanomnom commented on code/game/machinery/deployable.dm:
```diff
@@ -191,6 +191,27 @@
+	max_integrity = 100
+	proj_pass_rate = 25
+	armor = list("melee" = 10, "bullet" = 30, "laser" = 70, "energy" = 50, "bomb" = 10, "bio" = 100, "rad" = 100, "fire" = 10, "acid" = 0)
+	deploy_time = 10
```
```suggestion
	deploy_time = 1 SECONDS
```

ninjanomnom commented on code/game/machinery/deployable.dm:
```diff
@@ -191,6 +191,27 @@
+/obj/structure/barricade/security/holobarricade/Initialize()
+	. = ..()
+	addtimer(CALLBACK(src, .proc/deploy), deploy_time)
+	QDEL_IN(src, 250)
```
```suggestion
	QDEL_IN(src, 25 SECONDS)
```

ninjanomnom commented on code/modules/projectiles/guns/energy/laser.dm:
```diff
@@ -138,3 +139,68 @@
+
+/obj/item/external_lens/afterattack(atom/movable/AM, mob/user, flag)
+	. = ..()
+	if(user && istype(AM, restricted_type) && AM.type == restricted_type)
```
`istype(AM, restricted_type) && AM.type == restricted_type`
The istype accomplishes what you want here, as it is this makes it so it cant be applied to subtypes

Tlaltecuhtli replied:
subtypes were affected with only istype()

Tlaltecuhtli committed:
klhyf

Tlaltecuhtli committed:
klhyf

optimumtact commented:
there shouldn't need to be this much remedial work tlal

ExcessiveUseOfCobblestone commented on code/modules/projectiles/boxes_magazines/external_lens.dm:
```diff
@@ -0,0 +1,63 @@
+
+/obj/item/external_lens/afterattack(atom/movable/AM, mob/user, flag)
+	. = ..()
+	if(user && istype(AM, restricted_type) && AM.type == restricted_type)
```
the second and the third type are duplicates with the second covering more than the third.

ninjanomnom commented on code/datums/components/laser_components.dm:
```diff
@@ -0,0 +1,44 @@
+/datum/component/extralasers/Initialize(_ammo, _lens_path)
+	if(!istype(parent, /obj/item/gun/energy/laser))
+		return COMPONENT_INCOMPATIBLE
+	attach(_ammo, _lens_path)
```
Move this to RegisterWithParent

ninjanomnom commented on code/datums/components/laser_components.dm:
```diff
@@ -0,0 +1,44 @@
+
+/datum/component/extralasers/UnregisterFromParent()
+	UnregisterSignal(parent, COMSIG_PARENT_ATTACKBY)
+
```
Detach in UnregisterFromParent

ninjanomnom commented on code/datums/components/laser_components.dm:
```diff
@@ -0,0 +1,44 @@
+/datum/component/extralasers
+	dupe_mode = COMPONENT_DUPE_UNIQUE_PASSARGS
+	var/lens_path
+	var/obj/item/ammo_casing/energy/ammo
```
Clean this up in a Destroy proc

Dorsisdwarf commented:
Just like Paranoia

Tlaltecuhtli committed:
u

Tlaltecuhtli committed:
u

Tlaltecuhtli committed:
clean

Tlaltecuhtli committed:
clean

ninjanomnom commented on code/datums/components/laser_components.dm:
```diff
@@ -0,0 +1,48 @@
+	RegisterWithParent()
+
+/datum/component/extralasers/Destroy()
+	. = ..()
```
Destroy is different in that you need to call parent last

ninjanomnom commented on code/datums/components/laser_components.dm:
```diff
@@ -0,0 +1,48 @@
+
+/datum/component/extralasers/InheritComponent(datum/newcomp, orig, list/arglist)
+	. = ..()
+	UnregisterFromParent()
```
Don't call this yourself

ninjanomnom commented on code/datums/components/laser_components.dm:
```diff
@@ -0,0 +1,48 @@
+	var/turf/T = get_turf(parent)
+	new lens_path(T)
+
+	ammo = arglist[1]
```
Don't you need to new it?

ninjanomnom commented on code/datums/components/laser_components.dm:
```diff
@@ -0,0 +1,48 @@
+
+	ammo = arglist[1]
+	lens_path = arglist[2]
+	RegisterWithParent()
```
Don't call this yourself

ninjanomnom commented on code/datums/components/laser_components.dm:
```diff
@@ -0,0 +1,48 @@
+	L.update_icon(TRUE)
+
+/datum/component/extralasers/proc/attackby(datum/source, obj/item/I, mob/user, params)
+	. = ..()
```
You just made this proc, there is no parent to call

ninjanomnom commented on code/datums/components/laser_components.dm:
```diff
@@ -0,0 +1,48 @@
+	UnregisterSignal(parent, COMSIG_PARENT_ATTACKBY)
+	var/obj/item/gun/energy/laser/L = parent
+	if(L.chambered)
+		L.chambered = null
```
I was intending for you to just move the proc call for attaching/detaching it

ninjanomnom commented on code/modules/projectiles/boxes_magazines/external_lens.dm:
```diff
@@ -0,0 +1,63 @@
+
+/obj/item/external_lens/afterattack(atom/movable/AM, mob/user, flag)
+	. = ..()
+	if(user && AM.type == restricted_type)
```
```suggestion
	if(user && istype(AM, restricted_type))
```

Tlaltecuhtli committed:
v

Tlaltecuhtli committed:
v

Tlaltecuhtli commented:
istype checks dont work here is proof:
![weed(1)](https://user-images.githubusercontent.com/33834933/60946635-f5b97900-a2ee-11e9-84e9-d83c3a2764f9.png)
while the .type check does work

Tlaltecuhtli committed:
Merge branch 'master' into laserthing

ninjanomnom commented:
I have no clue what that's supposedly proving. Do you only want it to be used on the plain laser gun type and not subtypes?

ninjanomnom commented on code/modules/research/techweb/all_nodes.dm:
```diff
@@ -1,1056 +1,1054 @@
-//Current rate: 135000 research points in 90 minutes
-
-//Base Nodes
-/datum/techweb_node/base
```
Looks like you messed up line endings in this file

ninjanomnom commented on code/datums/components/laser_components.dm:
```diff
@@ -0,0 +1,52 @@
+
+/datum/component/extralasers/Destroy()
+	qdel(ammo)
+	..()
```
```suggestion
	return ..()
```

Tlaltecuhtli commented:
well yes it was supposed to be plain laser gun only, maybe i can give it a blacklist to prevent stuff like weaponized laser tags and have any gun with them

Tlaltecuhtli committed:
technodes

Tlaltecuhtli committed:
technodes

Tlaltecuhtli committed:
w

Tlaltecuhtli committed:
Merge branch 'master' into laserthing

Tlaltecuhtli committed:
size

Tlaltecuhtli committed:
size

ninjanomnom commented:
Alright the type check is kinda fine for that then

ninjanomnom commented on code/datums/components/laser_components.dm:
```diff
@@ -0,0 +1,52 @@
+
+/datum/component/extralasers/Destroy()
+	qdel(ammo)
+	return ..()
```
You need to call detach in destroy

While you're at it please move this proc up in the file. It's generally a bit easier to read when core procs like Initialize/Destroy etc are higher up near to the type definition

ninjanomnom commented on code/modules/cargo/packs.dm:
```diff
@@ -279,6 +279,33 @@
 					/obj/item/clothing/suit/armor/vest)
 	crate_name = "armor crate"
 
+/datum/supply_pack/security/assorted_lens
```
Remove this, as cobby said it's sidestepping needing the research which is a bad plan

wesoda25 commented:
Merge merge merge!

Tlaltecuhtli committed:
v

Tlaltecuhtli committed:
v

Tlaltecuhtli committed:
Update packs.dm

Tlaltecuhtli committed:
Update packs.dm

Tlaltecuhtli committed:
techweb update

Tlaltecuhtli committed:
techweb update

Tlaltecuhtli committed:
Merge branch 'master' into laserthing

Tlaltecuhtli committed:
adds 1 space

Tlaltecuhtli commented:
i text compared it with the master file and the only difference is the designs i added not sure why github says 1000 line changes

nemvar committed:
Replaces the slime extract in the clown biodome with a superlube sprayer. (#45134)

tgstation-server committed:
Automatic changelog generation for PR #45134 [ci skip]

Akrilla committed:
Species important for life (#45179)

* important to live items

* outfit

* autodoc

* comment change

* return

tgstation-server committed:
Automatic changelog generation for PR #45179 [ci skip]

Akrilla committed:
[READY] Contractor Hub (#45117)

* Contract Hub + Pinpointer + more contracts

* Show rep

* Better errors

* Dynamic hub

* Bug fixes

* Show items

* No html in desc

* compile

* contractor support

* limit

* allow limits of 1

web edit

* Equip and send partner + turf helper

* working reinforcements

* Cleanup

* removing debug

* blackout

* test

* Working buttons

* no debug

* prelight

* finalized interface

* better fit

* hub icon

* conflict

* text

* error

* kit tweaks

* items

* fixup

* fixup

* bugfix

* ui fix

tgstation-server committed:
Automatic changelog generation for PR #45117 [ci skip]

Dennok committed:
baseturf_helper fix (#44937)

* Update mapping_helpers.dm

* up

* up

* mod baseturf_helper

to have baseturf_to_replace = /turf/baseturf_bottom

* ups comma

* up

* Revert "up"

This reverts commit 7fb8a2e74543cce6f20cf4f9f1a30ba5de23068c.

* up

* Revert "up"

This reverts commit e2e3db5a8b4f59db349e29f2491e815235236842.

* Revert "Revert "up""

This reverts commit eff83a900d6b9e8b3bacb51ff7432e751a01ac07.

* Revert "up"

This reverts commit 7fb8a2e74543cce6f20cf4f9f1a30ba5de23068c.

* Revert "ups comma"

This reverts commit 5829dd96e8169be015897e38a525e86459c1844e.

* Revert "mod baseturf_helper"

This reverts commit f5a1c7f399ba3f285cdf18b1975be9f1516a2575.

* Update mapping_helpers.dm

* Update mapping_helpers.dm

* Update mapping_helpers.dm

* Update mapping_helpers.dm

tgstation-server committed:
Automatic changelog generation for PR #44937 [ci skip]

ninjanomnom commented:
It's because you changed the character used to mark the end of the line

bobbah committed:
fixed surgery being broken how did nobody notice this LOL (#45182)

tgstation-server committed:
Automatic changelog generation for PR #45182 [ci skip]

kriskog committed:
Sentient disease cures consistently harder, fixes uncurable diseases (#45148)

* Sentient disease cures consistently harder.

* Some comments.

* comment update.

tgstation-server committed:
Automatic changelog generation for PR #45148 [ci skip]

MrFluffster committed:
Mining Base Changes: Less empty office, more stuff, stuff that used to be is back (#45166)

* Changes to Mining Base to make it better, brings back the old storage closets

* Yes

* Changes lockers to locked, labels the cameras correctly

* does what cabby wanted

* fixed an oopsie

tgstation-server committed:
Automatic changelog generation for PR #45166 [ci skip]

Commit:
Improving upon the Metastation bar design (#45169)

* Ok Work Redone

* Airalarm & Firelock moved, Windows added, newscaster added

* Puts the preperation table in a better place

* Adds a extra table and swaps for r-tables

tgstation-server committed:
Automatic changelog generation for PR #45169 [ci skip]

AnturK committed:
Fixes MMI's hearing cult chat sometimes. (#45187)

* Fixes MMI's hearing deadchat somtimes.

* Alright, less safe but also less expensive safety net.

Gamer025 committed:
Fixed supply shuttle safety warning (#45164)

The safety warning preventing players from reaching centcom did not include any mention of cyborgs or the AI which however still trigger the warning.

tgstation-server committed:
Automatic changelog generation for PR #45164 [ci skip]

MMMiracles committed:
Gateway exiling counts for assassination/mutiny objectives (#45120)

* exiling someone via gateway now counts properly for mutiny/assassination objectives

* papers please

* ninja changes

* removes redundant else, istype

tgstation-server committed:
Automatic changelog generation for PR #45120 [ci skip]

tgstation-server committed:
Automatic changelog compile, [ci skip]

ninjanomnom committed:
Adds SHOULD_CALL_PARENT to Destroy() (#45188)

Timers, signals, and components need to know when the thing they're 
attached to is getting destroyed or bad things happen. Since any datum 
can make use of these every destroy needs to call parent.

ninjanomnom committed:
Add autodocs and linting tags to elements (#45189)

actioninja committed:
fixes an incorrect area on donut

Mickyan committed:
Random pixel shift for dropped/thrown objects (#45155)

* pixels

* offset for any throw, critical bedsheet bugfix

* nyeh

* might as well (autodoc)

* renamed proc

* removed redundant src

* moved call from dropped() to dropItemToGround()

* handles storage dumping, customizable max offset

* adjusted values

* inline

tgstation-server committed:
Automatic changelog generation for PR #45155 [ci skip]

tgstation-server committed:
Automatic changelog compile, [ci skip]

Fikou committed:
Clown Mutation (Clumsy) and Wizard Mutations can no longer be cured. (#45201)

* no mutadone cure

* no loop

* oops

* mutation fix

tgstation-server committed:
Automatic changelog generation for PR #45201 [ci skip]

tgstation-server committed:
Automatic changelog compile, [ci skip]

Dennok committed:
Fix hitting unpowered ORM (#45203)

Now unpowered ore redemption machine can be attacked.

tgstation-server committed:
Automatic changelog generation for PR #45203 [ci skip]

Gamer025 committed:
Singlecard description change (#45205)

Changed the description for a single (playing) card to be more accurate.

tgstation-server committed:
Automatic changelog generation for PR #45205 [ci skip]

optimumtact committed:
Ratvar callback no longer fires if item deleted by damage taken (#45215)

Tlaltecuhtli committed:
add line ender

ShizCalev committed:
Merge pull request #45223 from actioninja/donutareafix

fixes an incorrect area on donut

tgstation-server committed:
Automatic changelog generation for PR #45223 [ci skip]

nemvar committed:
Fixes (#45210)

tgstation-server committed:
Automatic changelog generation for PR #45210 [ci skip]

kriskog committed:
Medipen and regen cores can now be used in hand (#45161)

* Medipen and regen cores can now be activated in hand.

* Autodoc

* Old regen core description.

* Cyborg proc begone

* cleanup, no tk shenanigans.

* Removed unusable pen.

* better return.

tgstation-server committed:
Automatic changelog generation for PR #45161 [ci skip]

Tlaltecuhtli committed:
will it blend

Tlaltecuhtli committed:
Merge branch 'laserthing' of https://github.com/Tlaltecuhtli/tgstation into laserthing

Tlaltecuhtli commented:
new pr fixes the techweb problem and all the rest
